cmd/machine-os-builder/defualts.go

```diff
@@ -0,0 +1,170 @@
+			Namespace: ctrlcommon.MCONamespace,
+		},
+		Data: globalPullSecret.Data,
+		Type: globalPullSecret.Type,
+	}
+
+	_, err = kubeclient.CoreV1().Secrets(ctrlcommon.MCONamespace).Create(ctx, secretCopy, metav1.CreateOptions{})
+	return ignoreAlreadyExistsErr(err)
+}
+
+// Creates an OS ImageStream within the MCO namespace to push the final OS image to.
+func createOSImagestream(ctx context.Context, cb *clients.Builder) error {
+	is := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "os-image",
+			Namespace: ctrlcommon.MCONamespace,
+		},
+		Spec: imagev1.ImageStreamSpec{},
+	}
+
+	_, err := cb.ImageClientOrDie(componentName).ImageV1().ImageStreams(ctrlcommon.MCONamespace).Create(ctx, is, metav1.CreateOptions{})
+	return ignoreAlreadyExistsErr(err)
+}
+
+// Gets the secret name for a secret within the MCO namespace that can push to
+// the os-image ImageStream.
+func getBuilderPushSecretName(ctx context.Context, cb *clients.Builder) (string, error) {
+	secrets, err := cb.KubeClientOrDie(componentName).CoreV1().Secrets(ctrlcommon.MCONamespace).List(ctx, metav1.ListOptions{})
+
+	if err != nil {
+		return "", err
+	}
+
+	names := []string{}
+	for _, secret := range secrets.Items {
+		names = append(names, secret.Name)
+		if strings.HasPrefix(secret.Name, "builder-dockercfg") {
+			return secret.Name, nil
+		}
+	}
+
+	return "", fmt.Errorf("builder push secret name not found, found: %v", names)
+}
+
+// Determines if we should set up the defaults based upon our CLI flags.
+func getOrCreateBuildControllerConfigMap(ctx context.Context, cb *clients.Builder) (*corev1.ConfigMap, error) {
+	cmName := fmt.Sprintf("%s/%s", ctrlcommon.MCONamespace, build.OnClusterBuildConfigMapName)
+
+	kubeclient := cb.KubeClientOrDie(componentName)
+
+	onClusterBuildConfigMap, err := kubeclient.CoreV1().ConfigMaps(ctrlcommon.MCONamespace).Get(ctx, build.OnClusterBuildConfigMapName, metav1.GetOptions{})
+	if err == nil {
+		klog.Infof("ConfigMap %s found, will use", cmName)
+		return onClusterBuildConfigMap, nil
+	}
+
+	// If this isn't an IsNotFound error, return it to the caller to deal with.
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Below this point, it is assumed that we have an IsNotFound error.
+
+	// If we've opted to create defaults, let's do that now.
+	if startOpts.createDefaults {
+		klog.Infof("ConfigMap %q not found and --create-defaults set, will attempt to create / set defaults:", cmName)
+		return createDefaults(ctx, cb)
+	}
+
+	// If we've opted not to create defaults, return the noopErr which we look
+	// for later so we can enter the no-op loop.
+	return nil, errFoo
+}
+
+// Copies the global pull secret into the MCO namespace, creates an
+// ImageStream, identifies the builder push secret, and creates a ConfigMap
+// with all of these default values so that the BuildController can function
+// out of the box.
+func createDefaults(ctx context.Context, cb *clients.Builder) (*corev1.ConfigMap, error) {
```

Comment:
I'm reluctant to keep the `createDefaults()` functionality in its current form. Here's why:
1. The setup process for on-cluster builds is a one-time process that a cluster admin would be expected to perform before opting into on-cluster builds.
2. Aside from the operator portion of the MCO, no other MCO components do any kind of default setup.
3. We'll need to modify the deployment manifest to start the binary with the `--create-defaults` flag.
4. While the above is a simple change, it will cause the `machine-os-builder` to perform these actions every time it starts up. That could cause problems when the `machine-os-builder` pod restarts.

Instead, what I think we should do is simplify it thusly:
1. Check if the `on-cluster-build-config` ConfigMap exists. If it exists, continue on to start the BuildController as-is. If not, return an error.
2. Drop all references to the builder secret, ImageStream, and global pull secret.